Rework WebPageProxy::creationParameters function parameters

squelart opened the pull request:
#### 229250e6cd7fa24eab4423c217c814d977dc60e7
<pre>
Rework WebPageProxy::creationParameters function parameters
<a href="https://bugs.webkit.org/show_bug.cgi?id=269823">https://bugs.webkit.org/show_bug.cgi?id=269823</a>
<a href="https://rdar.apple.com/116201784">rdar://116201784</a>

Reviewed by Alex Christensen.

Two calls to WebPageProxy::creationParameters() modify the parameters before
using them in messages. These calls are now forced to instead provide all
the necessary information at once, and the returned WebPageCreationParameters
shouldn&apos;t need any further modifications.

* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
The SubframeProcessPageParameters struct definition has been moved outside
of WebPageCreationParameters, so that a forward declaration is enough when
declaring WebPageProxy::creationParameters.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::creationParametersForProvisionalPage):
(WebKit::WebPageProxy::creationParametersForRemotePage):
Added new function parameters, as needed for provisional and remote pages,
and helper functions that only take the minimum required parameters.

* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
Using the updated WebPageProxy::creationParameters function.
Since it doesn&apos;t require doctoring, it can be used directly in the
message creation.
Also, the call may be delayed until &amp; if it is actually needed.

* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
Using the updated WebPageProxy::creationParameters function.
Since it doesn&apos;t require doctoring, it can be used directly in the
message creation.

Canonical link: <a href="https://commits.webkit.org/275251@main">https://commits.webkit.org/275251@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d0ac707267a654daa8b3c9f8412c6915ad1d8601

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41176 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20189 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43554 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43739 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37268 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23324 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17520 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34075 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41750 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17181 "Found 1 new test failure: fast/forms/ios/file-upload-panel.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35462 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14720 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14912 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45060 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37428 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36779 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40527 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15991 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13111 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38906 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/41223 "Build is in progress. Recent messages:") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17610 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17662 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5513 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17254 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1573,6 +1574,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void didCancelForOpenPanel();
 
     WebPageCreationParameters creationParameters(WebProcessProxy&, DrawingAreaProxy&, RefPtr<API::WebsitePolicies>&& = nullptr);
```
Instead of adding two more functions that each call creationParameters then modify the result, I was thinking that it would be better to just add some more parameters to creationParameters:

WebPageCreationParameters creationParameters(WebProcessProxy&, DrawingAreaProxy&, bool isProcessSwap, RefPtr<API::WebsitePolicies>&& = nullptr, std::optional<SubframeProcessPageParameters>&& = std::nullopt, etc.);

That way the function shows all of what the result depends on and returns a complete result instead of a caller wondering which function to call and what needs to be done with the result after calling the function.

squelart replied:
My rationale included:
- The new functions were clearly intended for prov/remote pages.
- The new functions only took whatever data was required.
- The new functions did return a complete result, ready for message-sending.
- The function calls were as short as possible.

Now we have one giant function declaration, where callers need to explicitly provide extra information (that was implied by the ...ForX function names), and to skip function parameters that are not relevant (but only in one case so far, so it's not too bad.)

Please have a look and confirm that that's what you prefer, thanks.

Note: I had to pass SubframeProcessPageParameters by raw pointer, to avoid having to `#include WebPageCreationParameters.h` in WebPageProxy.h; Is there a better way?
Otherwise if we were to add the include, we would be able to pass `std::optional<WebPageCreationParameters::SubframeProcessPageParameters>&&` instead.

achristensen07 replied:
I do want all functions to take in all the parameters they need and return an object that needs no further modification.  Having a creationParametersForProvisionalPage and creationParametersForRemotePage and maybe another wrapper would be useful, though.  Also, all parameters like "true" should be given a name.

squelart replied:
A compromise! Much appreciated.

squelart committed:
Rework WebPageProxy::creationParameters function parameters
https://bugs.webkit.org/show_bug.cgi?id=269823
rdar://116201784

Reviewed by Alex Christensen.

Two calls to WebPageProxy::creationParameters() modify the parameters before
using them in messages. These calls are now forced to instead provide all
the necessary information at once, and the returned WebPageCreationParameters
shouldn't need any further modifications.

* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
The SubframeProcessPageParameters struct definition has been moved outside
of WebPageCreationParameters, so that a forward declaration is enough when
declaring WebPageProxy::creationParameters.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::creationParametersForProvisionalPage):
(WebKit::WebPageProxy::creationParametersForRemotePage):
Added new function parameters, as needed for provisional and remote pages,
and helper functions that only take the minimum required parameters.

* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
Using the updated WebPageProxy::creationParameters function.
Since it doesn't require doctoring, it can be used directly in the
message creation.
Also, the call may be delayed until & if it is actually needed.

* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
Using the updated WebPageProxy::creationParameters function.
Since it doesn't require doctoring, it can be used directly in the
message creation.

Canonical link: https://commits.webkit.org/275251@main
